lib/dap/filter/geoip2.rb

```diff
@@ -0,0 +1,157 @@
+require 'maxmind/db'
+
+module Dap
+module Filter
+
+require 'dap/utils/misc'
+
+module GeoIP2Library
+  GEOIP_DIRS = [
+    File.expand_path( File.join( File.dirname(__FILE__), "..", "..", "..", "data")),
+    "/var/lib/geoip",
+    "/var/lib/geoip2"
+  ]
+  GEOIP_CITY = %W{ GeoLite2-City.mmdb }
+  GEOIP_ASN = %W{ GeoLite2-ASN.mmdb }
+
+  @@geo_city = nil
+  @@geo_asn = nil
+
+  GEOIP2_CITY_DATABASE_PATH = ENV["GEOIP2_CITY_DATABASE_PATH"]
+  GEOIP2_ASN_DATABASE_PATH = ENV["GEOIP2_ASN_DATABASE_PATH"]
+
+  if GEOIP2_CITY_DATABASE_PATH
+    if ::File.exist?(GEOIP2_CITY_DATABASE_PATH)
+      @@geo_city = MaxMind::DB.new(GEOIP2_CITY_DATABASE_PATH, mode: MaxMind::DB::MODE_MEMORY)
+    end
+  else
+    GEOIP_DIRS.each do |d|
+      GEOIP_CITY.each do |f|
+        path = File.join(d, f)
+        if ::File.exist?(path)
+          @@geo_city = MaxMind::DB.new(path, mode: MaxMind::DB::MODE_MEMORY)
+          break
+        end
+      end
+    end
+  end
+
+  if GEOIP2_ASN_DATABASE_PATH
+    if ::File.exist?(GEOIP2_ASN_DATABASE_PATH)
+      @@geo_asn = MaxMind::DB.new(GEOIP2_ASN_DATABASE_PATH, mode: MaxMind::DB::MODE_MEMORY)
+    end
+  else
+    GEOIP_DIRS.each do |d|
+      GEOIP_ASN.each do |f|
+        path = File.join(d, f)
+        if ::File.exist?(path)
+          @@geo_asn = MaxMind::DB.new(path, mode: MaxMind::DB::MODE_MEMORY)
+          break
+        end
+      end
+    end
+  end
+end
+
+
+#
+# Add GeoIP2 tags using the MaxMind GeoIP2::City
+#
+class FilterGeoIP2City
+  include BaseDecoder
+  include GeoIP2Library
+
+  GEOIP2_LANGUAGE = ENV["GEOIP2_LANGUAGE"] || "en"
+  LOCALE_SPECIFIC_NAMES = %w(city.names continent.names country.names registered_country.names represented_country.names)
+  DESIRED_GEOIP2_KEYS = %w(
+    city.geoname_id
+    continent.code continent.geoname_id
+    country.geoname_id country.iso_code country.is_in_european_union
+    location.accuracy_radius location.latitude location.longitude location.metro_code location.time_zone
+    postal.code
+    registered_country.geoname_id registered_country.iso_code registered_country.is_in_european_union
+    represented_country.geoname_id represented_country.iso_code represented_country.is_in_european_union represented_country.type
+    traits.is_anonymous_proxy traits.is_satellite_provider
+  )
+
+  attr_reader :locale_specific_names
+  def initialize(args={})
+    @locale_specific_names = LOCALE_SPECIFIC_NAMES.map { |lsn| "#{lsn}.#{GEOIP2_LANGUAGE}" }
+    super
+  end
+
+  def decode(ip)
+    unless @@geo_city
+      raise "No MaxMind GeoIP2::City data found"
+    end
+    return unless (geo_hash = @@geo_city.get(ip))
+    ret = defaults
+
+    if geo_hash.include?("subdivisions")
+      # handle countries that are divided into various subdivisions.  generally 1, sometimes 2
+      subdivisions = geo_hash["subdivisions"]
+      geo_hash.delete("subdivisions")
+      ret["geoip2.city.subdivisions.length"] = subdivisions.size.to_s
```

Comment:
So I forgot to bring this up -- but what do you think of this model? Specifically, having this filter generate an _entirely flat_ structure that looks like:

```
{
...
"geoip2.city.subdivisions.length": "0",
"geoip2.city.subdivisions.0.geoname_id": "1234",
"geoip2.city.subdivisions.0.iso_code": "23456"
...
}
```

vs

```
{
...
"geoip2.city.subdivisions": [
  {
    "geoname_id": "1234",
    "iso_code": "23456",
    ...
   }
...
```

IOW, using an array for subdivisions vs the flat structure? I used the flat structure with godap with the assumption that, in general, dap should strive to only generate flat structures from its filters that generate new fields, and the only way arrays should be introduced is from the source data itself, or explicitly as the result of a filter operation that is documented to generate them ("`_array_`" filters).

FWIW, I agree with the model here (godap follows this obviously, I assume that's where you took insipration from), but I think it's worth discussing.

Reply:
I think you may have your suggestions swapped.  What godap does in your WIP PR and what this WIP dap PR does wrt subdivisions is the same -- flat.   Arrays can be useful but to your point they should only come directly from the input data or as a the result of a conversion.